File: algorithms/code/leetcode/lc042_trapping_rain_water/lc042_trapping_rain_water.java
package leetcode.lc042_trapping_rain_water;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Comparator;

/**
 * https://leetcode.com/problems/trapping-rain-water/ #hard
 */
public final class LC042TrappingRainWater {
    private static class MappingComparator implements Comparator<Integer>, Serializable {
        private int[] other;

        MappingComparator(final int[] otherValue) {
            other = otherValue;
        }

        public int compare(final Integer i1, final Integer i2) {
            return other[i2] - other[i1];
        }
    }

    private int amount(final int[] height, final int from, final int to) {
        int amount = Math.min(height[from], height[to]) * (to - from - 1);
        for (int i = from + 1; i < to; i++) {
            amount -= height[i];
        }
        return amount;
    }

    public int trap(final int[] height) {
        if (height == null || height.length < 3) {
            return 0;
        }
        // Sort the terrain
        Integer[] sorted = new Integer[height.length];
        for (int i = 0; i < height.length; i++) {
            sorted[i] = i;
        }
        Arrays.sort(sorted, new MappingComparator(height));
        // Fill from the top
        // (pick highest and then extend "exclusion zone")
        int count = 0;
        int left = sorted[0];
        int right = sorted[0];
        for (int i : sorted) {
            if (right < i) {
                count += amount(height, right, i);
                right = i;
            }
            if (i < left) {
                count += amount(height, i, left);
                left = i;
            }
        }
        return count;
    }
}
package leetcode.lc042_trapping_rain_water;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/trapping-rain-water/ #hard
 */
public class LC042TrappingRainWaterTests {
    @Test
    public void testExample() throws Exception {
        int[] terrain = { 0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1 };
        assertEquals(6, new LC042TrappingRainWater().trap(terrain));
    }

    @Test
    public void testNothing() throws Exception {
        assertEquals(0, new LC042TrappingRainWater().trap(null));
        assertEquals(0, new LC042TrappingRainWater().trap(new int[] { 0, 1 }));
    }
}
```rust
// Define a struct to compare two integers based on the height array.
#[derive(Debug)]
struct MappingComparator {
    other: [i32; height.len()],
}

impl MappingComparator {
    // Constructor to initialize the comparator with the given height array.
    fn new(other_value: &[i32]) -> Self {
        let mut cmp = MappingComparator { other: [0; other_value.len()] };
        cmp.other.clone_from_slice(other_value);
        cmp
    }

    // Compare two indices based on their corresponding heights in the 'other' array.
    fn compare(&self, i1: usize, i2: usize) -> std::cmp::Ordering {
        self.other[i2] - self.other[i1]
    }
}

// Define a function to calculate the amount of water that can be trapped between two indices.
fn amount(height: &[i32], from: usize, to: usize) -> i32 {
    let min_height = height[from].min(height[to]);
    (to - from - 1) as i32 * min_height
        + height.iter().skip(from + 1).take_while(|&&x| x <= min_height).sum::<i32>()
}

// Define a function to calculate the total amount of water that can be trapped in the terrain.
fn trap(height: &[i32]) -> i32 {
    if height.is_empty() || height.len() < 3 {
        return 0;
    }

    // Sort the indices based on their corresponding heights in the 'other' array.
    let mut sorted_indices = Vec::new();
    for i in 0..height.len() {
        sorted_indices.push(i);
    }
    sorted_indices.sort_by(|a, b| MappingComparator::new(&height).compare(b, a));

    // Initialize variables to keep track of the left and right boundaries.
    let mut left = sorted_indices[0];
    let mut right = sorted_indices[0];

    // Iterate through the sorted indices to find the trapped water areas.
    let mut count = 0;
    for i in sorted_indices {
        if right < i {
            count += amount(height, left, i);
            left = i;
        }
        if i < left {
            count += amount(height, i, right);
            right = i;
        }
    }

    count
}

// Define tests for the 'trap' function.
#[cfg(test)]
mod tests {
    use super::*;

    // Test case 1: Trapping rain water in a terrain with multiple peaks and valleys.
    #[test]
    fn test_example() -> Result<(), Box<dyn std::error::Error>> {
        let height = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
        assert_eq!(trap(&height), 6);
        Ok(())
    }

    // Test case 2: Trapping rain water in an empty terrain.
    #[test]
    fn test_nothing() -> Result<(), Box<dyn std::error::Error>> {
        let height = &[0, 1];
        assert_eq!(trap(height), 0);
        Ok(())
    }
}
```

File: algorithms/code/leetcode/lc042_trapping_rain_water/lc042_trapping_rain_water.rs
// https://leetcode.com/problems/trapping-rain-water/
// #hard

use std::cmp::Ordering;

pub struct Solution;

impl Solution {
    fn mapping_comparator(i1: &usize, i2: &usize, map: &[i32]) -> Ordering {
        if map[*i1] > map[*i2] {
            return Ordering::Less;
        }
        if map[*i1] < map[*i2] {
            return Ordering::Greater;
        }
        Ordering::Equal
    }

    fn amount(height: &[i32], from: usize, to: usize) -> i32 {
        let amount = height[from].min(height[to]) * (to - from - 1) as i32;
        amount - height[from + 1..to].iter().sum::<i32>()
    }

    pub fn trap(height: &[i32]) -> i32 {
        if height.len() < 3 {
            return 0;
        }
        // Sort the terrain
        let mut sorted = Vec::new();
        for i in 0..height.len() {
            sorted.push(i);
        }
        sorted.sort_by(|i1, i2| Self::mapping_comparator(i1, i2, height));
        // Fill from the top
        // (pick highest and then extend "exclusion zone")
        let mut count = 0;
        let mut left = sorted[0];
        let mut right = sorted[0];
        for i in sorted {
            if right < i {
                count += Self::amount(height, right, i);
                right = i;
            }
            if i < left {
                count += Self::amount(height, i, left);
                left = i;
            }
        }
        count
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let terrain = vec![0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
        assert_eq!(Solution::trap(&terrain), 6);
    }

    #[test]
    fn test_nothing() {
        assert_eq!(Solution::trap(&vec![0, 1]), 0);
    }

    #[test]
    fn test_comparator() {
        let height = vec![0i32, 1i32];
        assert_eq!(
            Solution::mapping_comparator(&1usize, &0usize, &height),
            Ordering::Less
        );
        assert_eq!(
            Solution::mapping_comparator(&0usize, &1usize, &height),
            Ordering::Greater
        );
        assert_eq!(
            Solution::mapping_comparator(&0usize, &0usize, &height),
            Ordering::Equal
        );
    }
}
